Client/Frontend/Browser/TabManager.swift

```diff
@@ -232,30 +232,70 @@ class TabManager : NSObject {
 }
 
 extension TabManager {
+    class SavedTab: NSObject, NSCoding {
+        let url: NSURL?
+        let isReaderMode: Bool
+        let isSelected: Bool
+
+        init(url: NSURL, isReaderMode: Bool, isSelected: Bool) {
+            self.url = url
+            self.isReaderMode = isReaderMode
+            self.isSelected = isSelected
+        }
+
+        required init(coder: NSCoder) {
+            self.url = coder.decodeObjectForKey("url") as? NSURL
+            self.isReaderMode = coder.decodeBoolForKey("isReaderMode")
+            self.isSelected = coder.decodeBoolForKey("isSelected")
+        }
+
+        func encodeWithCoder(coder: NSCoder) {
+            coder.encodeObject(url, forKey: "url")
+            coder.encodeBool(isReaderMode, forKey: "isReaderMode")
+            coder.encodeBool(isReaderMode, forKey: "isSelected")
+        }
+    }
+
     func encodeRestorableStateWithCoder(coder: NSCoder) {
-        coder.encodeInteger(count, forKey: "tabCount")
-        coder.encodeInteger(selectedIndex, forKey: "selectedIndex")
-        for i in 0..<count {
-            let tab = tabs[i]
-            coder.encodeObject(tab.url!, forKey: "tab-\(i)-url")
+        var savedTabs = [SavedTab]()
+        for (tabIndex, tab) in enumerate(tabs) {
+            if let url = tab.url, displayURL = tab.displayURL {
+                let savedTab = SavedTab(url: displayURL, isReaderMode: ReaderModeUtils.isReaderModeURL(url), isSelected: tabIndex == selectedIndex)
```

Comment:
Alternative suggestion that might be a bit more robust: check for all URLs starting with "http://localhost:<port>" and replace it with some special token (about:localhost?). Then, when we restore, replace this token with "http://localhost:<port>" again, with the new port.

A couple advantages to this method:
 1. We don't need special flags (`isReaderMode`, `isErrorPage`, etc) for different local page types. That means #466 won't need to make any changes after this lands.
 1. We could use the tab's actual URL instead of the display URL. That means special cases, like about:home mentioned above, will "just work".

What do you think?